include/dlaf/multiplication/general/impl.h

```diff
@@ -19,13 +19,52 @@
 #include <dlaf/communication/communicator_grid.h>
 #include <dlaf/matrix/distribution.h>
 #include <dlaf/matrix/index.h>
+#include <dlaf/matrix/matrix.h>
+#include <dlaf/matrix/matrix_ref.h>
 #include <dlaf/matrix/panel.h>
 #include <dlaf/multiplication/general/api.h>
 #include <dlaf/sender/when_all_lift.h>
+#include <dlaf/util_matrix.h>
 
 namespace dlaf::multiplication {
 namespace internal {
 
+template <Backend B, Device D, class T>
+void General<B, D, T>::callNN(const blas::Op opA, const blas::Op opB, const T alpha,
+                              MatrixRef<const T, D>& mat_a, MatrixRef<const T, D>& mat_b, const T beta,
+                              MatrixRef<T, D>& mat_c) {
+  namespace ex = pika::execution::experimental;
+
+  if (opA != blas::Op::NoTrans || opB != blas::Op::NoTrans)
+    DLAF_UNIMPLEMENTED(opA, opB);
+
+  using matrix::multipliable_sizes;
+  DLAF_ASSERT_HEAVY(multipliable_sizes(mat_a.size(), mat_b.size(), mat_c.size(), opA, opB),
+                    "Multiplication incompatible matrix sizes.", mat_a.size(), mat_b.size(),
+                    mat_c.size(), opA, opB);
+  DLAF_ASSERT_HEAVY(multipliable_sizes(mat_a.blockSize(), mat_b.blockSize(), mat_c.blockSize(), opA,
+                                       opB),
+                    "Multiplication incompatible tile sizes.");
+  DLAF_ASSERT_HEAVY(mat_c.size().isEmpty() ||
+                        multipliable_sizes(mat_a.distribution().tileSize({0, 0}),
+                                           mat_b.distribution().tileSize({0, 0}),
+                                           mat_c.distribution().tileSize({0, 0}), opA, opB),
+                    "Multiplication incompatible tile sizes in first row/col. "
+                    "(Are you using a matrix with offset not aligned with tile?)");
```

Comment:
Do we want to have an helper and move preconditions there as we did for `GeneralSub`?

https://github.com/eth-cscs/DLA-Future/blob/3097ab4f06c8fb2c94ea804c75698727de1ce82f/include/dlaf/multiplication/general.h#L60-L74